### PR TITLE
WMDP-209: Add run-task script to make running ECS tasks manually easier

### PR DESCRIPTION
**Pull request by chouinar**
## Ticket

https://wicmtdp.atlassian.net/browse/WMDP-209

## Changes
Adds the `run-task.sh` script which is a wrapper around the `aws ecs run-task ..` CLI command which helps setup a lot of the configurations.

## Context for reviewers
There are a lot of things needed to run an ECS task, most notably:
* The task you want to run
* The subnet & security groups you want to use
* Who is running the task
* What container to run the task in
* What environment you are running in

Some of those like the task and environment should be manual each time, but things like the subnet and security group should ideally be pretty automated with the ability to override them, which is what this script does.

The most complex bit of this is probably the bash script commands run. Most notably the `jq` command which is used to parse and process values into JSON formatted data using a mixture of CLI and template files. Happy to add more examples of exactly how that works if the existing explanation isn't perfect.

Note that because we're still building out some of the infra, a few things are more hardcoded than would be ideal, but I've created a TODO in the readme to help keep aware of those.

## Testing
`./bin/run-ecs-task/run-task.sh -a your.name -e test -t test-csv-handler-definition` will startup the ECS task and run the CSV generation script.

[Cloudwatch Logs](https://us-east-1.console.aws.amazon.com/cloudwatch/home?region=us-east-1#logsV2:log-groups/log-group/mock-api/log-events/csv-handler$252Ftest-mock-api-container$252F65f3d208a91349a5addff9e8b55aa388) for an example run - although by default it creates the CSV file locally so it disappears after the run finishes.

I've also run this same command with the `command` field overriden in the container overrides to be `["poetry", "run", "db-migrate-up"]` in order to do the DB migrations.

### CODE REVIEW

File: app/bin/run-ecs-task/README.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+It also heavily uses `jq` for parsing and generating the CSV configurations.
+
+# Usage
+
+`./run-task.sh -a your.name -e test -t your-task-definition` is the most basic way to run a task.
+
+The script has the following parameters:
+* `-a` - The author running the script (eg. if your name is Bob Smith, put `-a bob.smith`)
+* `-e` - The environment you want to run the script in (eg. `-e test`)
```
Comment (rocketnova) [inline app/bin/run-ecs-task/README.md:17]: Clarifying question: This should match the environment as named in terraform?

File: app/bin/run-ecs-task/network_config.json.tpl [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "awsvpcConfiguration": {
+    "assignPublicIp": "ENABLED",
+    "securityGroups": [],
+    "subnets": []
```
Comment (rocketnova) [inline app/bin/run-ecs-task/network_config.json.tpl]: Is this where the defaults should be set once we have them?

File: app/bin/run-ecs-task/run-task.sh [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+fi
+
+# If no security groups were specified, fetch them from AWS. The response is formatted as
+# {
+#   "SecurityGroups": [
+#       {"GroupName": "csv-handler", GroupId: "sg-123456789", ...}  
+# ]}
+if [ -z "$SECURITY_GROUPS" ]; then
+    # The jq command grabs all returned security group IDs and formats it into a JSON array
+    SECURITY_GROUPS=$(aws ec2 describe-security-groups --group-names csv-handler | jq -r --compact-output "[.SecurityGroups[].GroupId]")
```
Comment (rocketnova) [inline app/bin/run-ecs-task/run-task.sh:63]: Could these `aws` commands fail if the user has to specify a `--profile` flag when running aws CLI based on their custom setup?

File: app/bin/run-ecs-task/run-task.sh [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    "--started-by=$AUTHOR"
+    "--task-definition=$TASK_NAME"
+    "--launch-type=FARGATE"
+    "--platform-version=1.4.0"
+    --network-configuration "$NETWORK_CONFIG"
+    --overrides "$OVERRIDES"
+    )
+
+# Print out the arguments
+printf " ... %s" "${AWS_ARGS[@]}"
```
Comment (rocketnova) [inline app/bin/run-ecs-task/run-task.sh:108]: Could this contain secrets at all? Such as from the environment variables in the overrides json?

File: app/bin/run-ecs-task/README.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+* `-g` - (Optional) - The security group IDs to attach to the ECS task (eg. `-g '["sg-1234", "sg-3456"]'`)
+    * If no security groups are specified, the `csv-handler` security group is fetched (TODO)
+
+## Template Files
+Two JSON template files exist which are used to help format the JSON configurations.
+
+### Network config
+The network config should be left as-is, as the subnet and security groups are specified as described in the usage section above.
+
+### Container overrides
```
Comment (rocketnova) [inline app/bin/run-ecs-task/README.md:32]: Could you add some more description here of how one should use the container override json? Is the idea that I would edit the file if I wanted to change the default override?

**Review by rocketnova — COMMENTED** [review]
Could you include some of this in your README.md? It's very good and should not only exist in the PR description:

```
There are a lot of things needed to run an ECS task, most notably:

    The task you want to run
    The subnet & security groups you want to use
    Who is running the task
    What container to run the task in
    What environment you are running in

Some of those like the task and environment should be manual each time, but things like the subnet and security group should ideally be pretty automated with the ability to override them, which is what this script does.

The most complex bit of this is probably the bash script commands run. Most notably the jq command which is used to parse and process values into JSON formatted data using a mixture of CLI and template files. 
```

File: app/bin/run-ecs-task/network_config.json.tpl (hunk 2 above)
Comment (chouinar) [reply app/bin/run-ecs-task/network_config.json.tpl]: No, anything written in securityGroups or subnets gets 100% overriden by the script as written (either based on CLI inputs or whats found in AWS). Defaults would go in the commands that query AWS with whatever security group name(s) we default to + the VPC you want to run in to get all of its subnets.

File: app/bin/run-ecs-task/README.md (hunk 1 above)
Comment (chouinar) [reply app/bin/run-ecs-task/README.md:17]: ~Yes. It currently is only used to specify the ECS cluster to run in (which is named exactly the environment like `test`). If we had a bit more setup, we'd likely use it for a few other pieces like the name of the container. Actually, let me just change that so it's not hardcoded to `test-mock-api-container` but instead does `{env}-mock-api-container`

File: app/bin/run-ecs-task/run-task.sh (hunk 3 above)
Comment (chouinar) [reply app/bin/run-ecs-task/run-task.sh:63]: They would, you need a working AWS cli + configuration to get this working. Let me find the docs that explain that and add it to the readme.

File: app/bin/run-ecs-task/run-task.sh (hunk 4 above)
Comment (chouinar) [reply app/bin/run-ecs-task/run-task.sh:108]: Technically yes, although would that be an issue for someone running locally if they just manually entered them?

I don't think you can even override parameter store env variables due to how they get loaded into the container (at runtime?) - or at least I remember running into that issue before.

File: app/bin/run-ecs-task/README.md (hunk 5 above)
Comment (chouinar) [reply app/bin/run-ecs-task/README.md:32]: Yes, that's right. I can add more detail.

File: app/bin/run-ecs-task/network_config.json.tpl (hunk 2 above)
Comment (chouinar) [reply app/bin/run-ecs-task/network_config.json.tpl]: To make this more clear, I'm setting the "default" value that'll be overriden to something saying to go look at the readme.

**Comment by chouinar** [comment]
> Could you include some of this in your README.md? It's very good and should not only exist in the PR description:

I do have it within the options section, should I make that clearer somehow? Just updated the PR - let me know if there's anywhere more you want detail.

**Review by rocketnova — APPROVED** [review]
LGTM! Thank you!

**Pull request merged**